### PR TITLE
[CONTRIBUTING.md] Add section about merging ECS conventions

### LINKED ISSUE #329: Provide guidance when there is conflict with ECS conventions
When adopting ECS conventions following open-telemetry/oteps/pull/222, there are certain cases where ECS conventions could be improved in some way. For example a given ECS convention may suffer from any of the following problems:
1. It may not follow existing OTel semantic conventions guidelines (e.g. plural vs singular guidelines).
2. Its namespacing is too lax for the usual OTel standards (e.g. maybe the convention name could be a namespace instead).
3. It could be too general and more specific conventions would be needed to allow for certain use-cases.
4. There could be overlap with existing OTel conventions

Three specific cases where this issues arise are as follows:
- On #66 I proposed adopting semantic conventions related to kernels. ECS already has `os.kernel` but this could be a namespace, it has a full `uname` output that would need to be manually parsed
- On #131 both `host.ip` and `host.mac` don't follow the usual plural guidelines and IP could be a namespace for IP protocol related conventions.
- On #181, `url.domain` overlaps with `server.address` significantly

I would like to have a document with specific guidance as to what to do when we have these conflicting requirements between compatibility with ECS and existing guidelines/good practices. Some questions I would like to have answered in the document are:

1. Should ECS conventions be amended to follow plural guidelines?
2. Can an ECS convention be used as namespace? If not, what should be done? (e.g., should the kernel version be `os.kernel_version` to avoid namespacing?)
3. Are redundancy or possible better names valid arguments when considering if renaming an ECS convention?

This relates to https://github.com/open-telemetry/semantic-conventions/pull/181#discussion_r1312729412

cc @AlexanderWert

### PR DESCRIPTION
**Pull request by mx-psi**
Fixes #329

## Changes

Add new section to CONTRIBUTING.md with expectations when merging existing ECS conventions.

## Merge requirement checklist

* [ ] [CONTRIBUTING.md](https://github.com/open-telemetry/semantic-conventions/blob/main/CONTRIBUTING.md) guidelines followed.
* [ ] [CHANGELOG.md](https://github.com/open-telemetry/semantic-conventions/blob/main/CHANGELOG.md) updated for non-trivial changes.
* [ ] [schema-next.yaml](https://github.com/open-telemetry/semantic-conventions/blob/main/schema-next.yaml) updated with changes to existing conventions.

### CODE REVIEW

**Review by kaiyan-sheng — APPROVED** [review]

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -180,4 +180,21 @@ make misspell-correction
...
 
+## Merging existing ECS conventions
+
+The Elastic Common Schema (ECS) is being merged into OpenTelemetry Semantic
+Conventions per [OTEP 222][otep222]. When adding a semantic convention that
+exists in some form in ECS, consider the following guidelines:
+
+- Prefer using the existing ECS name when possible. In particular:
+  - Do not amend ECS names to follow plural guidelines or other syntactical
+    attribute naming rules.
```
Comment (dmitryax) [inline CONTRIBUTING.md:191]: It's a bit unclear what are the "plural guidelines or other syntactical attribute naming rules".

Also, is this file the best place to define it? Maybe it should go to https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/semantic_conventions/README.md where all the guidelines are defined? It seems like we are scattering the guidelines across different places

Comment (ChrsMark) [reply CONTRIBUTING.md:191]: The linked file is metrics specific though, no?
I had found https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/common/attribute-naming.md#name-pluralization-guidelines about the pluralization rule which seems to be more generic?

Maybe https://github.com/open-telemetry/semantic-conventions/blob/main/docs/general/README.md is a more generic one for this repo. I'm not sure if it's searchable in the docs page though.

For now I would be fine with keeping it in the contribution guide as well.

**Review by ChrsMark — APPROVED** [review]

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -180,4 +180,21 @@ make misspell-correction
...
 
+## Merging existing ECS conventions
+
+The Elastic Common Schema (ECS) is being merged into OpenTelemetry Semantic
+Conventions per [OTEP 222][otep222]. When adding a semantic convention that
+exists in some form in ECS, consider the following guidelines:
+
+- Prefer using the existing ECS name when possible. In particular:
+  - Do not amend ECS names to follow plural guidelines or other soft 
+    recommendations.
```
Comment (trask) [inline CONTRIBUTING.md:191]: do we know how many of these there will be? if there's a lot, maybe we should just drop the pluralization guideline completely (#267)?

(not that i'm in favor of dropping the guideline, but if there's a lot of these then it feels like we will be essentially defacto dropping it)

Comment (mx-psi) [reply CONTRIBUTING.md:191]: I think we can handle the discussion on dropping the specific rule separately. The main motivation for this line is the discussion in #203, but the rule is also meant to apply to the [counter naming rules](https://github.com/open-telemetry/semantic-conventions/blob/main/docs/general/metrics.md#naming-rules-for-counters-and-updowncounters), the rules listed [here](https://github.com/open-telemetry/semantic-conventions/blob/main/docs/general/metrics.md#instrument-naming) and other rules that are not a 'MUST'. In general, the rule I am trying to write down here is "it is okay to ignore a 'SHOULD' rule when integrating existing ECS conventions for the sake of backwards compatibility with ECS".  I think these rules are still useful to have to guide contributions that do not have an equivalent in ECS.

Comment (AlexanderWert) [reply CONTRIBUTING.md:191]: > not that i'm in favor of dropping the guideline, but if there's a lot of these then it feels like we will be essentially defacto dropping it

I think having that guideline as a `soft` recommendation still makes sense for any new attributes and where there are no strong arguments against it.

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -180,4 +180,21 @@ make misspell-correction
   - Send staging tag as PR for review.
 - Create a tag `v{version}` on the merged PR and push remote.
 
+## Merging existing ECS conventions
+
+The Elastic Common Schema (ECS) is being merged into OpenTelemetry Semantic
+Conventions per [OTEP 222][otep222]. When adding a semantic convention that
+exists in some form in ECS, consider the following guidelines:
+
+- Prefer using the existing ECS name when possible. In particular:
```
Comment (joaopgrassi) [inline CONTRIBUTING.md:189]: Recently it was introduced container attributes from [OCI](https://github.com/open-telemetry/semantic-conventions/pull/159). What will be our guidelines when:

- Another standard (such OCI in this case) has attributes different from the ones from ECS? Will we still take ECS? Will ECS adapt to the other, potentially OSS standard?

Comment (ChrsMark) [reply CONTRIBUTING.md:189]: I think that in https://github.com/open-telemetry/semantic-conventions/pull/159 we tried to follow the OCI spec and not directly use any naming convention. In this there was no breaking-change trade off, right?

Comment (joaopgrassi) [reply CONTRIBUTING.md:189]: Yes, I just mentioned that as an example of another open standard that could clash with ECS.

Comment (AlexanderWert) [reply CONTRIBUTING.md:189]: @joaopgrassi I think that would be rather an edge case that we would discuss case by case anyways, WDYT?

Comment (mx-psi) [reply CONTRIBUTING.md:189]: I would take this as 'comparable evidence' on the rule for 'proposing a name that differs from the ECS convention'. Maybe we can explicitly add "prior work on a different standard" as a piece of evidence that can be 'used as evidence'?

Comment (pyohannes) [reply CONTRIBUTING.md:189]: In my understanding this case is well covered by this point:

> * If proposing a name that differs from the ECS convention, provide usage data, user issue reports or feature requests or comparable evidence about the alternatives.

If there's a "competing" standard, I think it needs to be evaluated on a case-by-case basis whether the additional benefits for users of being consistent with the "competing" standard outweigh the pain of breaking ECS.

Comment (mx-psi) [reply CONTRIBUTING.md:189]: Added this explicitly on that bullet point for clarity

File: CONTRIBUTING.md (hunk 1 above)
Comment (AlexanderWert) [reply CONTRIBUTING.md]: @dmitryax does that resolve your comment?

File: CONTRIBUTING.md [hunk 4]
```diff
@@ -197,4 +197,21 @@ to merge**.
...
+  - Do not amend ECS names to follow plural guidelines or other soft
```
Comment (joaopgrassi) [inline CONTRIBUTING.md:207]: I guess the goal of this sentence is to say something like: "Don't change the ECS names just to comply with OTel attribute guidelines" right? Like the pluralization rules.

In that case, what about we go with something more generic:


```suggestion
  - ECS names should not be modified to comply with OpenTelemetry general attribute guidelines. 
```

OR
```suggestion  
  - Special consideration must be applied for ECS attribute names that do not comply with the OpenTelemetry general attribute guidelines. 
```

OR

```suggestion
 - If an ECS attribute name does not comply with the OpenTelemetry general attribute guidelines and changing it is not desirable, provide data to support why changing the ECS attribute change cannot be done.
```

These are just suggestions.

**Review by joaopgrassi — COMMENTED** [review]
I don't really love the "follow plural guidelines or other soft recommendation" wording we have there. I think we can do better :)

File: CONTRIBUTING.md (hunk 4 above)
Comment (AlexanderWert) [reply CONTRIBUTING.md:207]: @joaopgrassi I like that direction!

How about?:

> ECS names should not be modified for the _only reason_ of complying with OpenTelemetry general attribute guidelines.

File: CONTRIBUTING.md (hunk 2 above)
Comment (mx-psi) [reply CONTRIBUTING.md]: @trask This has been reworded in the last iteration, PTAL

File: CONTRIBUTING.md [hunk 5]
```diff
@@ -197,5 +197,22 @@ to merge**.
...
 
+## Merging existing ECS conventions
+
+The Elastic Common Schema (ECS) is being merged into OpenTelemetry Semantic
+Conventions per [OTEP 222][otep222]. When adding a semantic convention that
+exists in some form in ECS, consider the following guidelines:
+
+- Prefer using the existing ECS name when possible. In particular:
+  - Do not modify an ECS name for the only reason of complying with
+    OpenTelemetry general attribute guidelines.
```
Comment (frzifus) [inline CONTRIBUTING.md:208]: Based on the discussions on `host.ip` we might want to define this a bit more.

Comment (mx-psi) [reply CONTRIBUTING.md:208]: Do you mean refining as in "we should make it clear what the general attribute guidelines are" or "specifying which attribute guidelines we should follow"?

Comment (joaopgrassi) [reply CONTRIBUTING.md:208]: with "general attribute guidelines" I was mostly referring to this: https://github.com/open-telemetry/semantic-conventions/blob/main/docs/general/metrics.md#general-guidelines. That is the central place we have such guidelines.. and if something is not there, we should add. (for ex #394 is not there yet)

Comment (mx-psi) [reply CONTRIBUTING.md:208]: The recommendations here https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/common/attribute-naming.md should also be part of this IMO. If you think it makes sense to explicitly list what guidelines we are talking about I can add an explicit list.

Comment (AlexanderWert) [reply CONTRIBUTING.md:208]: @joaopgrassi I think the guidelines @mx-psi linked (https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/common/attribute-naming.md) are more generic and (should) reference the one you linked.

https://github.com/open-telemetry/semantic-conventions/blob/main/docs/general/metrics.md#general-guidelines is only about metrics, right?

Comment (AlexanderWert) [inline CONTRIBUTING.md:208]:
```suggestion
    [OpenTelemetry general attribute naming guidelines](https://opentelemetry.io/docs/specs/otel/common/attribute-naming/).
```

Comment (joaopgrassi) [reply CONTRIBUTING.md:208]: Right, all of those apply, yes. I think it would be good to link them in here.

**Comment by mx-psi** [comment]
Trying to summarize the work needed here from previous SIGs and private discussions I have had about this. These are the questions that the community wants to understand to move this forward:

1. How many conventions from ECS don't follow our existing guidelines? So that we understand to how many misaligned conventions we are committing. Alexandra Conrad is working on this.
2. What tools/capabilities does Elastic have today to migrate their users if we were to rename conventions? 
3. What tools would it be feasible for Elastic to build? 
4. Is it harder to migrate users in some situations? For example, does `host.ip` being a leaf node in ECS and us introducing `host.ip.addresses` causes more trouble than a simple rename?

cc @AlexanderWert @MikePaquette

**Comment by trisch-me** [comment]
Hey @mx-psi I have finished my investigation on possible naming conflicts. Please check it below. 

TL;DR:

-  pluralisation, i.e. ECS has fields which are arrays but key is singular. As prominent examples of it - `*.ip` and `*.mac` for arrays of ip or mac addresses, ~40 examples found in ECS
- `*.bytes` fields which would be changed to the `*.size` ~15 examples
- `*.packets` fields, I think they might be a subject to change as with previous field  ~10 examples
- `@` character in name, there is only 1 such example (`@timestamp`), not sure if it against naming rules
- array of nested objects: ECS contains fields (for example [dns.answers](https://www.elastic.co/guide/en/ecs/current/ecs-dns.html#field-dns-answers)) which are arrays containing objects. ECS defines fields of this nested objects. I'm not sure yet how this would go into Otel schema. From this [data model](https://github.com/open-telemetry/opentelemetry-specification/blob/eda62e4baf906ac8225a9ee69fa97ff7f48dbccc/specification/logs/data-model.md#field-attributes) can we say that for attribute of type `map<string, any>`, `any` could be such an object too?


Here is detailed list of Namespaces and possible conflicting fields in it:

- Base: `@timestamp`
- Client: `client.bytes`, `client.packets`
- Container: `container.disk.(write|read).bytes`, `container.image.hash.all` (array of hashes), `container.image.tag` (array of image tags), `container.network.(ingress|egress).bytes`
- Destination: `destination.bytes`, `destination.packets`
- DNS: `dns.resolved_ip` (Array containing all IPs seen in answers.data)
- Email: `email.(bcc|cc|from|reply_to|to).address` - The email addresses
- Event:` event.category`,` event.type` - arrays
- Host: `host.disk.(write|read).bytes`,` host.ip` (Host ip addresses), `host.mac` (Host MAC addresses, array of values), `host.network.(ingress|egress).bytes`, `host.network.(ingress|egress).packets`
- Http: `http.request.body.bytes`, `http.request.bytes` (body + headers), `http.response.body.bytes`, `http.response.bytes`
- Network: `network.bytes`, `network.packets`
- Observer: `observer.ip`, `observer.mac` are arrays
- Orchestrator: `orchestrator.resource.annotation` (The list of annotations added to the resource.), `orchestrator.resource.ip`, `orchestrator.resource.label` - these are all arrays
- Process: `process.thread.capabilities.effective` (set of capabilities used by the kernel to perform permission checks for the thread), `process.thread.capabilities.permitted` - these are arrays and we have rule in metrics that namespaces shouldn't be pluralised, but I'm not sure about these cases
- Related: `related.hash`, `related.hosts`, `related.ip`, `related.user` - all arrays with data gathered all over the event, I'm not sure if it will be moved to Otel but potential conflict. (A concrete example is IP addresses, which can be under host, observer, source, destination, client, server, and network.forwarded_ip. If you append all IPs to related.ip, you can then search for a given IP trivially, no matter where it appeared, by querying related.ip:192.0.2.15.)
- Server: `server.bytes`, `server.packets`
- Source: `source.bytes`, `source.packets`
- Threat: `threat.(group|software).alias`, `threat.tactic.(id|name|reference)`, `threat.technique.(id|name|reference)`, `threat.technique.subtechnique.(id|name|reference)` - these are all arrays
- User: `user.id` is defined not as array but a user might have more than 1 ID associated
- Vulnerability: `vulnerability.category` is an array
- x509 Certificate: `x509.issuer.(common_name| country|locality|organization|organizational_unit|state_or_province)`, `x509.subject.(common_name| country|locality|organization|organizational_unit|state_or_province)` - these are all arrays

**Comment by pyohannes** [comment]
> @ character in name, there is only 1 such example (@timestamp), not sure if it against naming rules

Can we possibly ignore this conflict, as we likely won't have a `@timestamp` field on semantic conventions (as that's part data structure of signals themselves, and doesn't need to be in semantic conventions)?

**Comment by trisch-me** [comment]
> 1. How many conventions from ECS don't follow our existing guidelines? So that we understand to how many misaligned conventions we are committing. Alexandra Conrad is working on this.

This one has been answered above. Please also note, this might be not full list of all possible conflicts, while adding new fields we might find other naming conflicts, but I think this list is comprehensive enough

> 3. Is it harder to migrate users in some situations? For example, does host.ip being a leaf node in ECS and us introducing host.ip.addresses causes more trouble than a simple rename?

Yes, changing name is already a breaking change, but changing structure - in this case going from simple leaf with value/array of values to complex node with additional data underneath will have really strong negative impact on ECS users and tooling. I would strongly advise against it. Please also check comments in [this PR](https://github.com/open-telemetry/semantic-conventions/pull/203#issuecomment-1752866265) for additional context

File: CONTRIBUTING.md [hunk 6]
```diff
@@ -197,5 +197,22 @@ to merge**.
...
 
+## Merging existing ECS conventions
+
+The Elastic Common Schema (ECS) is being merged into OpenTelemetry Semantic
+Conventions per [OTEP 222][otep222]. When adding a semantic convention that
+exists in some form in ECS, consider the following guidelines:
+
+- Prefer using the existing ECS name when possible. In particular:
+  - Do not modify an ECS name for the only reason of complying with
+    [OpenTelemetry general attribute naming guidelines](https://opentelemetry.io/docs/specs/otel/common/attribute-naming/).
```
Comment (joaopgrassi) [inline CONTRIBUTING.md:208]: We discussed this PR in the working group meeting on Nov 27. The general feeling is that this bullet point is "too permissive" and pretty much opens up to disregard completely the guidelines in place, including the ones which are important (e.g. character set).

We discussed these options/alternatives/improvements

1. Adapt this bullet point to better indicate which guidelines shouldn't be applied, or adapt the text such as it is not so broad 
2. Remove this paragraph completely
3. Close this PR and deal with problematic/conflicting ECS attributes as they appear
4. Indicate here that new attributes should be checked on ECS first if they exist there before suggesting new attributes in OTel to avoid even more conflicts

The conclusion was we should aim to do a mix of 1, and 4

Comment (trisch-me) [reply CONTRIBUTING.md:208]: Great summary @joaopgrassi 

I think logic-wise it's better start with second bullet point 
```
-  If proposing a name that differs from the ECS convention, provide usage
    data, user issue reports, feature requests, examples of prior work on a
    different standard or comparable evidence about the alternatives.
```
And then add more clarification, because most of the problems coming from `Name Pluralization guidelines`, everything else from naming guidelines are the same for ECS as well
```
- When no suitable alternatives are provided, altering an ECS name solely 
  for the purpose of complying with [Name Pluralization guidelines] (https://opentelemetry.io/docs/specs/otel/common/attribute-naming/#name-pluralization-guidelines) 
  SHOULD BE avoided.
- Before proposing a new attribute, it's advisable to check ECS first to ensure 
  a similar attribute doesn't already exist. If a match is found, consider adopting it 
  into Otel to minimize potential conflicts.
```

Comment (trask) [reply CONTRIBUTING.md:208]: i'd suggest slightly more permissive still: `SHOULD BE avoided` -> `MAY BE avoided`

File: CONTRIBUTING.md (hunk 2 above)
Comment (mx-psi) [reply CONTRIBUTING.md]: I am marking this as solved since we rediscussed this on a discussion below.

File: CONTRIBUTING.md (hunk 1 above)
Comment (mx-psi) [reply CONTRIBUTING.md]: @dmitryax can we resolve this?

**Comment by trisch-me** [comment]
I have changed the status of the issue to get more attention from other approvers

**Review by joaopgrassi — APPROVED** [review]
I think it looks good now :)

File: CONTRIBUTING.md [hunk 7]
```diff
@@ -199,5 +199,23 @@ to merge**.
...
+The Elastic Common Schema (ECS) is being merged into OpenTelemetry Semantic
+Conventions per [OTEP 222][otep222]. When adding a semantic convention that
+exists in some form in ECS, consider the following guidelines:
+
+- Prefer using the existing ECS name when possible. In particular:
+  - If proposing a name that differs from the ECS convention, provide usage
+    data, user issue reports, feature requests, examples of prior work on a
+    different standard or comparable evidence about the alternatives.
+  - When no suitable alternatives are provided, altering an ECS name solely
+    for the purpose of complying with [Name Pluralization guidelines](https://opentelemetry.io/docs/specs/otel/common/attribute-naming/#name-pluralization-guidelines)
```
Comment (arminru) [inline CONTRIBUTING.md:213]:
```suggestion
    for the purpose of complying with [Name Pluralization guidelines](docs/general/attribute-naming.md#name-pluralization-guidelines)
```
These guidelines have moved into this repo here.

File: CONTRIBUTING.md [hunk 8]
```diff
@@ -199,5 +199,23 @@ to merge**.
...
+    MAY BE avoided.
+- Do not use an existing ECS name as a namespace. If the name must differ, use a
+  different namespace name to avoid clashes or avoid using the namespace
+  entirely.
```
Comment (arminru) [inline CONTRIBUTING.md:217]:
```suggestion
  entirely. See the [ECS field reference] for existing namespaces.
```

File: CONTRIBUTING.md [hunk 9]
```diff
@@ -199,5 +199,23 @@ to merge**.
...
+
 [nvm]: https://github.com/nvm-sh/nvm/blob/master/README.md#installing-and-updating
 [stability guarantees]: https://github.com/open-telemetry/opentelemetry-specification/blob/v1.26.0/specification/versioning-and-stability.md#semantic-conventions-stability
+[otep222]: https://github.com/open-telemetry/oteps/pull/222
```
Comment (arminru) [inline CONTRIBUTING.md:221]:
```suggestion
[otep222]: https://github.com/open-telemetry/oteps/pull/222
[ECS field reference]: https://www.elastic.co/guide/en/ecs/current/ecs-field-reference.html
```

File: CONTRIBUTING.md (hunk 6 above)
Comment (mx-psi) [reply CONTRIBUTING.md]: Resolving per approval from Joao :)

**Comment by mx-psi** [comment]
Anything missing here? Or is this good to merge?
